### PR TITLE
nansum reduction function

### PR DESCRIPTION
**Pull request by OwenArnold**
Listed under #1221 

* `nansum` for Variable, DataArray and Dataset
* `sum` for DataArray (all dims) and Dataset (all dims)

### CODE REVIEW

File: core/include/scipp/core/element/arithmetic.h [hunk 1]
```diff
@@ -22,8 +22,33 @@ constexpr auto add_inplace_types =
...
+                 using ElementType = std::decay_t<decltype(a)>;
+                 auto constexpr zero = []() {
+                   if constexpr (is_ValueAndVariance_v<ElementType>)
+                     return ElementType{0, 0};
+                   else
+                     return ElementType{0};
+                 }();
+                 using std::isnan;
+                 if (isnan(a))
+                   a = zero;
```
Comment (SimonHeybrock) [inline core/include/scipp/core/element/arithmetic.h:45]: Does `a = std::decay_t<decltype(a)>{};` work?

File: core/include/scipp/core/element/arithmetic.h [hunk 2]
```diff
@@ -22,8 +22,33 @@ constexpr auto add_inplace_types =
...
+                 if (isnan(b)) {
+                   a += zero;
+                   return;
+                 }
+                 a += b;
```
Comment (SimonHeybrock) [inline core/include/scipp/core/element/arithmetic.h:50]: Why not
```suggestion
                 if (!isnan(b))
                   a += b;
```
?

File: core/include/scipp/core/element/arithmetic.h [hunk 3]
```diff
@@ -22,8 +22,33 @@ constexpr auto add_inplace_types =
              std::tuple<float, int64_t>, std::tuple<float, int32_t>,
              std::tuple<int64_t, bool>>;
 
+constexpr auto add_inplace_nan_types =
+    arg_list<double, float, int64_t, int32_t, std::tuple<double, float>,
+             std::tuple<float, double>, std::tuple<int64_t, int32_t>,
+             std::tuple<int32_t, int64_t>, std::tuple<double, int64_t>,
+             std::tuple<double, int32_t>, std::tuple<float, int64_t>,
+             std::tuple<float, int32_t>, std::tuple<int64_t, bool>>;
```
Comment (SimonHeybrock) [inline core/include/scipp/core/element/arithmetic.h:30]: Can we avoid this and reuse `add_inplace_types`? I suppose the problem is the `isnan` does not exist for all those types? Maybe an `if constexpr` can help (maybe in a heper lambda)?
```cpp
if constexpr(std::is_floating_point<T>)
  return isnan(x);
else
  return false;
}
```

File: python/reduction.cpp [hunk 4]
```diff
@@ -61,6 +61,30 @@ template <class T> void bind_sum_out(py::module &m) {
       py::call_guard<py::gil_scoped_release>());
 }
 
+template <class T> void bind_nansum(py::module &m) {
+  if constexpr (std::is_same_v<T, Variable>) {
+    m.def(
+        "nansum",
+        [](const typename T::const_view_type &x) { return nansum(x); },
+        py::arg("x"), py::call_guard<py::gil_scoped_release>());
+  }
```
Comment (SimonHeybrock) [inline python/reduction.cpp:70]: Since you are touching the related files anyway: Maybe this is a good opportunity to also provide this overload for `DataArray`, and same for `sum` above?

File: variable/reduction.cpp [hunk 5]
```diff
@@ -55,20 +55,31 @@ void sum_impl(const VariableView &summed, const VariableConstView &var) {
   accumulate_in_place(summed, var, element::plus_equals);
 }
 
-Variable sum(const VariableConstView &var, const Dim dim) {
+void nansum_impl(const VariableView &summed, const VariableConstView &var) {
+  if (contains_events(var))
+    throw except::TypeError("`nansum` can only be used for dense data, use "
+                            "`flatten` for event data.");
```
Comment (SimonHeybrock) [inline variable/reduction.cpp:61]: Suggest to remove this, it will cause conflicts with the event-list removal.

File: core/include/scipp/core/element/arithmetic.h (hunk 1 above)
Comment (OwenArnold) [reply core/include/scipp/core/element/arithmetic.h:45]: Unfortunately not, because such constructor for ValueAndVariance, but I think it would be reasonable to add one.

File: core/include/scipp/core/element/arithmetic.h (hunk 3 above)
Comment (OwenArnold) [reply core/include/scipp/core/element/arithmetic.h:30]: Yes, that is exactly the problem - I didn't like having to define the sub list of types again.

I did consider a solution along the lines of what you propose, but it is also misleading without reading the implementation. It supposes that `add_inplace_nan_types` is going to support time point etc.

File: core/include/scipp/core/element/arithmetic.h (hunk 1 above)
Comment (SimonHeybrock) [reply core/include/scipp/core/element/arithmetic.h:45]: I would have thought the default constructor should exist (compiler generated?)?

Comment (OwenArnold) [reply core/include/scipp/core/element/arithmetic.h:45]: No because non-default constructors are defined. However I note that there is a constructor that should allow us to do the above. https://github.com/scipp/scipp/blob/master/core/include/scipp/core/value_and_variance.h#L28. i.e. 
`a = std::decay_t<decltype(a)>{0}` should work.

File: dataset/variable_reduction.cpp [hunk 6]
```diff
@@ -35,6 +35,22 @@ VariableView sum(const VariableConstView &var, const Dim dim,
   return sum(var, dim, out);
 }
 
+Variable nansum(const VariableConstView &var, const Dim dim,
+                const MasksConstView &masks) {
+  if (const auto mask_union = irreducible_mask(masks, dim)) {
+    return nansum(applyMask(var, mask_union), dim);
```
Comment (SimonHeybrock) [inline dataset/variable_reduction.cpp:41]: Where is `applyMask` defined?

File: dataset/dataset_operations_common.h [hunk 7]
```diff
@@ -117,6 +117,28 @@ DataArray apply_to_data_and_drop_dim(const DataArrayConstView &a, Func func,
...
+DataArray apply_to_data_and_drop_all_dims(const DataArrayConstView &a,
+                                          Func func) {
+  const auto &dimensions = a.data().dims();
+  DataArray out(a);
+  for (scipp::index i = 0; i < dimensions.ndim(); ++i) {
+    out = apply_to_data_and_drop_dim(out, func, dimensions.label(i), a.masks());
+  }
+  return out;
+}
+
```
Comment (SimonHeybrock) [inline dataset/dataset_operations_common.h:141]: Why is this required? Is it not possible to use `reduce_all_dims`?

File: dataset/variable_reduction.cpp (hunk 6 above)
Comment (SimonHeybrock) [reply dataset/variable_reduction.cpp:41]: Ignore this, I found it.

File: dataset/dataset_operations_common.h (hunk 7 above)
Comment (OwenArnold) [reply dataset/dataset_operations_common.h:141]: Yes the logic is almost identical and can be templated. This does not however handle masks https://github.com/scipp/scipp/blob/master/variable/operations_common.h#L30-L38

Comment (SimonHeybrock) [reply dataset/dataset_operations_common.h:141]: You don't need that? See how it is used: https://github.com/scipp/scipp/blob/53ed25a4778f6b5343b0dd177681b4803eefdae9/variable/reduction.cpp#L159-L162

Since `sum(data_array, dim)` exists you should be able to use the same. The masks are part of `data_array` and do not need special handling, I belive?

File: dataset/reduction.cpp [hunk 8]
```diff
@@ -12,8 +12,12 @@
 #include "../variable/operations_common.h"
 #include "dataset_operations_common.h"
 
+using namespace scipp::core;
 namespace scipp::dataset {
 
+DataArray sum(const DataArrayConstView &a) {
+  return reduce_all_dims(a, [&](auto &&... _) { return sum(_...); });
```
Comment (SimonHeybrock) [inline dataset/reduction.cpp:19]: What is this capturing?

File: dataset/reduction.cpp [hunk 9]
```diff
@@ -28,6 +32,32 @@ Dataset sum(const DatasetConstView &d, const Dim dim) {
       d, [](auto &&... _) { return sum(_...); }, dim);
 }
 
+Dataset sum(const DatasetConstView &d) {
+  return apply_to_items(d, [](auto &&... _) { return sum(_...); });
+}
+
+DataArray nansum(const DataArrayConstView &a) {
+  return reduce_all_dims(a, [&](auto &&... _) { return nansum(_...); });
```
Comment (SimonHeybrock) [inline dataset/reduction.cpp:40]: Capturing what?

**Review by SimonHeybrock — COMMENTED** [review]
Missing file?

File: core/include/scipp/core/reduction.h [hunk 10]
```diff
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+// Copyright (c) 2020 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Owen Arnold
```
Comment (SimonHeybrock) [inline core/include/scipp/core/reduction.h:4]: Suggest to either remove authors, or keep the main authors of the code in the file (such that other devs know who to ask about it).

Does this belong into `core/`? It can't really operate on any objects from this module? On the other hand, it does not depend on anything higher level? Is `common/` a better place (is including `dim.h` required?)?

File: dataset/reduction.cpp (hunk 9 above)
Comment (OwenArnold) [reply dataset/reduction.cpp:40]: Good "catch" yes, nothing.

File: core/include/scipp/core/reduction.h (hunk 10 above)
Comment (OwenArnold) [reply core/include/scipp/core/reduction.h:4]: Good point. I'll move to common.

**Review by SimonHeybrock — APPROVED** [review]

File: variable/reduction.cpp [hunk 11]
```diff
@@ -3,39 +3,51 @@
...
 using namespace scipp::core;
+using scipp::common::reduce_all_dims;
 
 namespace scipp::variable {
 
+// Workaround VS C7526 (undefined inline variable) with dtype<> in template.
+bool is_dtype_bool(const VariableConstView &var) {
+  return var.dtype() == dtype<bool>;
+}
+bool is_dtype_int64(const VariableConstView &var) {
```
Comment (OwenArnold) [inline variable/reduction.cpp:23]: Thanks for the fix @Matthew-Andrew. Is this int function necessary? I didn't see it complain about dtype<int> in the failed build.

Comment (Matthew-Andrew) [reply variable/reduction.cpp:23]: Locally, once I fixed the bool complaint I started seeing an identical error for the use of `d_type<Int64_t>`
